src/package.ps1

```diff
@@ -0,0 +1,485 @@
+		[Collections.Hashtable]$Pkg
+	)
+	return "$(GetAirpowerPath)\ref\$($Pkg.Package)$(if (-not $Pkg.Tag.Latest) { "-$($Pkg.Tag | AsTagString)" })"
+}
+
+function ResolveRemoteRef {
+	param (
+		[Parameter(Mandatory, ValueFromPipeline)]
+		[Collections.Hashtable]$Pkg
+	)
+	$remote = GetRemoteTags
+	if (-not $remote.$($Pkg.Package)) {
+		throw "no such package: $($Pkg.Package)"
+	}
+	$want = $Pkg.Tag
+	foreach ($got in $remote.$($Pkg.Package)) {
+		$eq = $true
+		if ($null -ne $want.Major) {
+			$eq = $eq -and $want.Major -eq $got.Major
+		}
+		if ($null -ne $want.Minor) {
+			$eq = $eq -and $want.Minor -eq $got.Minor
+		}
+		if ($null -ne $want.Patch) {
+			$eq = $eq -and $want.Patch -eq $got.Patch
+		}
+		if ($null -ne $want.Build) {
+			$eq = $eq -and $want.Build -eq $got.Build
+		}
+		if ($eq) {
+			return "$($Pkg.Package)-$(($got.ToString()).Replace('+', '_'))"
+		}
+	}
+	throw "no such $($Pkg.Package) tag: $($Pkg.Tag)"
+}
+
+function GetLocalPackages {
+	$db = GetPwrDB
+	$pkgs = @()
+	foreach ($pkg in $db.pkgdb.keys) {
+		foreach ($tag in $db.pkgdb.$pkg.keys) {
+			$t = [Tag]::new($tag)
+			$digest = if ($t.None) { $tag } else { $db.pkgdb.$pkg.$tag }
+			$pkgs += [PSCustomObject]@{
+				Package = $pkg
+				Tag = $t
+				Digest = $digest | AsDigest
+				Size = $db.metadatadb.$digest.size | AsSize
+				# Signers
+			}
+		}
+	}
+	if (-not $pkgs) {
+		$pkgs = ,[PSCustomObject]@{
+			Package = $null
+			Tag = $null
+			Digest = $null
+			Size = $null
+		}
+	}
+	return $pkgs
+}
+
+function ResolvePackageDigest {
+	param (
+		[Parameter(Mandatory, ValueFromPipeline)]
+		[Collections.Hashtable]$Pkg
+	)
+	$db = GetPwrDB
+	return $db.pkgdb.$($Pkg.Package).$($Pkg.Tag | AsTagString)
+}
+
+function InstallPackage { # $db, $status
+	param (
+		[Parameter(Mandatory, ValueFromPipeline)]
+		[Collections.Hashtable]$Pkg
+	)
+	$db = GetPwrDB
+	$digest = $Pkg.Digest
+	$name = $Pkg.Package
+	$tag = $Pkg.Tag | AsTagString
+	if ($null -ne $db.pkgdb.$name.$tag -and $digest -ne $db.pkgdb.$name.$tag) {
+		$status = 'newer'
+		$old = $db.pkgdb.$name.$tag
+		$db.pkgdb.$name.Remove($tag)
+		$db.pkgdb.$name.$old = $null
+		if ($db.metadatadb.$old.refcount -gt 0) {
+			$db.metadatadb.$old.refcount -= 1
+		}
+	}
+	if ($null -eq $db.pkgdb.$name.$tag) {
+		if ($db.metadatadb.$digest) {
+			if ($db.pkgdb.$name.ContainsKey($digest)) {
+				$db.pkgdb.$name.Remove($digest)
+			}
+			$status = 'tag'
+			$db.metadatadb.$digest.refcount += 1
+		} else {
+			$status = 'new'
+			$db.metadatadb.$digest = @{
+				RefCount = 1
+				Size = $Pkg.Size
+			}
+		}
+		$db.pkgdb.$name += @{
+			"$tag" = $digest
+		}
+	} else {
+		$status = 'uptodate'
+	}
+	return $db, $status
+}
+
+function PullPackage {
+	param (
+		[Parameter(Mandatory, ValueFromPipeline)]
+		[Collections.Hashtable]$Pkg
+	)
+	$manifest = $Pkg | ResolveRemoteRef | GetManifest
+	$Pkg.Digest = $manifest | GetDigest
+	$Pkg.Size = $manifest | GetSize
+	WriteHost "$($pkg.Tag | AsTagString): Pulling $($Pkg.Package)"
```

Comment:
I realize this somewhat matches what `docker pull` prints, but it results in really odd looking strings like `8: Pulling jdk`. Could we change this to something like:
```suggestion
	WriteHost "Pulling $($Pkg.Package):$($pkg.Tag | AsTagString)"
```

Reply:
Yeah